src/instruction_graph_generator.cc

```diff
@@ -0,0 +1,2038 @@
+		// if scalar_box is up to date in that memory, it (the single element) must also be contiguous
+		auto& source_allocation = buffer.memories[source_mid].get_contiguous_allocation(scalar_reduction_box);
+
+		local_gather_copy_instr = create<copy_instruction>(command_batch, source_allocation.aid, gather_aid + m_local_nid * buffer.elem_size,
+		    source_allocation.box, scalar_reduction_box, scalar_reduction_box, buffer.elem_size,
+		    [&](const auto& record_debug_info) { record_debug_info(copy_instruction_record::copy_origin::gather, bid, buffer.debug_name); });
+		add_dependency(local_gather_copy_instr, fill_identity_instr, instruction_dependency_origin::write_to_allocation);
+		perform_concurrent_read_from_allocation(local_gather_copy_instr, source_allocation, scalar_reduction_box);
+	}
+
+	// 4. Gather remote contributions to the partial result array
+
+	const transfer_id trid(gather->consumer_tid, bid, gather->rid);
+	const auto gather_recv_instr = create<gather_receive_instruction>(command_batch, trid, gather_aid, node_chunk_size,
+	    [&](const auto& record_debug_info) { record_debug_info(buffer.debug_name, gather->gather_box, m_num_nodes); });
+	add_dependency(gather_recv_instr, fill_identity_instr, instruction_dependency_origin::write_to_allocation);
+
+	// 5. Perform the global reduction on the host by reading the array of inputs from the gather space and writing to the buffer's host allocation that covers
+	// `scalar_reduction_box`.
+
+	allocate_contiguously(command_batch, bid, host_memory_id, {scalar_reduction_box});
+
+	auto& host_memory = buffer.memories[host_memory_id];
+	auto& dest_allocation = host_memory.get_contiguous_allocation(scalar_reduction_box);
+
+	const auto reduce_instr = create<reduce_instruction>(command_batch, rid, gather_aid, m_num_nodes, dest_allocation.aid, [&](const auto& record_debug_info) {
+		record_debug_info(rcmd.get_cid(), bid, buffer.debug_name, scalar_reduction_box, reduce_instruction_record::reduction_scope::global);
+	});
+	add_dependency(reduce_instr, gather_recv_instr, instruction_dependency_origin::read_from_allocation);
+	if(local_gather_copy_instr != nullptr) { add_dependency(reduce_instr, local_gather_copy_instr, instruction_dependency_origin::read_from_allocation); }
+	perform_atomic_write_to_allocation(reduce_instr, dest_allocation, scalar_reduction_box);
+	buffer.track_original_write(scalar_reduction_box, reduce_instr, host_memory_id);
+
+	// 6. Free the gather space
+
+	const auto gather_free_instr = create<free_instruction>(
+	    command_batch, gather_aid, [&](const auto& record_debug_info) { record_debug_info(m_num_nodes * node_chunk_size, std::nullopt); });
+	add_dependency(gather_free_instr, reduce_instr, instruction_dependency_origin::allocation_lifetime);
+
+	buffer.pending_gathers.clear();
+
+	// The associated `runtime_reduction` info will be garbage-collected form the executor as we pass the reduction id on via the instruction_garbage member of
```

Comment:
It will once [the runtime-integration of reductions](https://github.com/celerity/celerity-runtime/blob/73075a19f11297344bfa6e47668bcfe92e0f1244/include/reduction.h) is merged.